### PR TITLE
Integrated the kyc section in finance tab

### LINKED ISSUE #140: Redesign: /Finance/KYC
**NOTICE** 

**Remember to checkout and send a pull request to the [redesign](https://github.com/openMF/mobile-wallet/tree/redesign)  branch.**

---

We are working on the app redesign. The prototype is done and can be found here: [show prototype](https://xd.adobe.com/view/8581e34d-8b91-49ec-7c21-000d98c8f18b-f3b0/?hints=off), and source files here: [show repository](https://github.com/letelete/mifos-pay-app-design). We will be working on the redesign implementation now.

This issue is about implementing the design and features of the KYC tab in the Finance section.

That's what we want to achieve:

image

### PR DESCRIPTION
**Pull request by manddy**
Fixes #140 

Please Add Screenshots If any UI changes.
kyc

### CODE REVIEW

**Comment by ankurs287** [comment]
@manddy Your branch has some conflicts. Kindly fix them and resend the PR.

**Comment by manddy** [comment]
@ankurs287  I have fixed all the conflicts kindly have a look

File: mifospay/gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -0,0 +1,5 @@
+distributionBase=GRADLE_USER_HOME
```
Comment (letelete) [inline mifospay/gradle/wrapper/gradle-wrapper.properties]: Why did you modify gradle-wrapper.properties

File: mifospay/gradlew [hunk 2]
```diff
@@ -0,0 +1,172 @@
+#!/usr/bin/env sh
```
Comment (letelete) [inline mifospay/gradlew]: Why did you modify gradlew

File: mifospay/gradlew.bat [hunk 3]
```diff
@@ -0,0 +1,84 @@
+@if "%DEBUG%" == "" @echo off
```
Comment (letelete) [inline mifospay/gradlew.bat]: Why did you modify gradlew.bat

File: mifospay/local.properties [hunk 4]
```diff
@@ -0,0 +1,8 @@
+## This file must *NOT* be checked into Version Control Systems,
```
Comment (letelete) [inline mifospay/local.properties]: Why did you modify local.properties

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/finance/ui/KycFragment.java [hunk 5]
```diff
@@ -1,19 +1,24 @@
...
+import org.mifos.mobilewallet.mifospay.kyc.ui.KYCDescriptionFragment;
 
 public class KycFragment extends BaseFragment {
 
     @Override
     public View onCreateView(LayoutInflater inflater, ViewGroup container,
                              Bundle savedInstanceState) {
-        View rootView = inflater.inflate(R.layout.fragment_kyc, container, false);
-        return rootView;
+        // Inflate the layout for this fragment
```
Comment (letelete) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/finance/ui/KycFragment.java]: This comment is redundant

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java [hunk 6]
```diff
@@ -37,6 +41,24 @@
...
+
+    @BindView(R.id.first)
+    LinearLayout f;
+
+    @BindView(R.id.second)
+    LinearLayout s;
+
+    @BindView(R.id.third)
+    LinearLayout t;
+
```
Comment (letelete) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: What are those, excuse me? It's not your local project, but global open-source app where you must keep the code style to the adopted standards

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java [hunk 7]
```diff
@@ -55,6 +77,24 @@
...
+    @BindView(R.id.completed1)
+    TextView completed1;
+
+    @BindView(R.id.completed2)
+    TextView completed2;
+
+    @BindView(R.id.completed3)
+    TextView completed3;
+
```
Comment (letelete) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: ...

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java [hunk 8]
```diff
@@ -95,20 +133,32 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
 
     @OnClick(R.id.cv_lvl1)
     public void onLevel1Clicked() {
-        replaceFragmentUsingFragmentManager(KYCLevel1Fragment.newInstance(), true,
-                R.id.container);
+        KYCLevel1Fragment kycLevel1Fragment=new KYCLevel1Fragment();
```
Comment (letelete) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: So many

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java [hunk 9]
```diff
@@ -95,20 +133,32 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+        fragmentTransaction.addToBackStack(null);
+        fragmentTransaction.commit();
+
     }
 
     @OnClick(R.id.cv_lvl2)
     public void onLevel2Clicked() {
-        replaceFragmentUsingFragmentManager(KYCLevel2Fragment.newInstance(), true,
-                R.id.container);
+        KYCLevel2Fragment kycLevel2Fragment=new KYCLevel2Fragment();
```
Comment (letelete) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: code

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java [hunk 10]
```diff
@@ -95,20 +133,32 @@ public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container,
...
+        fragmentTransaction.commit();
+
     }
 
     @OnClick(R.id.cv_lvl3)
     public void onLevel3Clicked() {
-        replaceFragmentUsingFragmentManager(KYCLevel3Fragment.newInstance(), true,
-                R.id.container);
+
+        KYCLevel3Fragment kycLevel3Fragment=new KYCLevel3Fragment();
```
Comment (letelete) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: repeats...

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java [hunk 11]
```diff
@@ -119,15 +169,24 @@ public void onFetchLevelSuccess(KYCLevel1Details kycLevel1Details) {
 
         if (currentLevel >= 1) {
             cvLvl1.setClickable(false);
-            llLvl1.setBackgroundResource(R.drawable.cardview_round_green);
+            text1.setTextColor(getResources().getColor(R.color.colorofkyvlevelsbutton));
```
Comment (letelete) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: Redundant code repeats

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java [hunk 12]
```diff
@@ -119,15 +169,24 @@ public void onFetchLevelSuccess(KYCLevel1Details kycLevel1Details) {
...
+            completed1.setText("Completed");
+            img1.setImageResource(R.drawable.ic_tick);
+            llLvl1.setBackgroundResource(R.drawable.cardview_round_white);
         }
         if (currentLevel >= 2) {
             cvLvl2.setClickable(false);
-            llLvl2.setBackgroundResource(R.drawable.cardview_round_green);
+            text2.setTextColor(getResources().getColor(R.color.colorofkyvlevelsbutton));
```
Comment (letelete) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: Redundant code repeats

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java [hunk 13]
```diff
@@ -119,15 +169,24 @@ public void onFetchLevelSuccess(KYCLevel1Details kycLevel1Details) {
...
+            completed2.setText("Completed");
+            img2.setImageResource(R.drawable.ic_tick);
+            llLvl2.setBackgroundResource(R.drawable.cardview_round_white);
         }
         if (currentLevel >= 3) {
             cvLvl3.setClickable(false);
-            llLvl3.setBackgroundResource(R.drawable.cardview_round_green);
+            text3.setTextColor(getResources().getColor(R.color.colorofkyvlevelsbutton));
+            completed3.setText("Completed");
```
Comment (letelete) [inline mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: Redundant code repeats

File: mifospay/src/main/res/layout/fragment_kyc_desc.xml [hunk 14]
```diff
@@ -1,132 +1,230 @@
...
+                            android:id="@+id/lvl2text"
+                            android:layout_width="wrap_content"
+                            android:layout_height="wrap_content"
+                            android:layout_margin="@dimen/value_10dp"
+                            android:text="@string/level_2"
+                            android:textAlignment="center"
+                            android:textColor="@color/white"
+                            android:textSize="@dimen/value_16sp" />
+
+                        <!--<TextView-->
```
Comment (letelete) [inline mifospay/src/main/res/layout/fragment_kyc_desc.xml]: Why is this here

File: mifospay/src/main/res/layout/fragment_kyc_desc.xml [hunk 15]
```diff
@@ -1,132 +1,230 @@
...
+                            android:id="@+id/lvl3text"
+                            android:layout_width="wrap_content"
+                            android:layout_height="wrap_content"
+                            android:layout_margin="@dimen/value_10dp"
+                            android:text="@string/level_3"
+                            android:textAlignment="center"
+                            android:textColor="@color/white"
+                            android:textSize="@dimen/value_16sp" />
+
+                        <!--<TextView-->
```
Comment (letelete) [inline mifospay/src/main/res/layout/fragment_kyc_desc.xml]: Why is this here

File: mifospay/src/main/res/values/colors.xml [hunk 16]
```diff
@@ -1,6 +1,6 @@
 <?xml version="1.0" encoding="utf-8"?>
 <resources>
-
+    <color name="colorbackground">#FAFAFA</color>
```
Comment (letelete) [inline mifospay/src/main/res/values/colors.xml]: colorBackground of... what? It's a redundant attribute. Just use `colorAccent`

File: mifospay/src/main/res/values/colors.xml [hunk 17]
```diff
@@ -37,7 +37,7 @@
 
     <color name="colorBottomViewNotFocusedIcon">@color/colorAccentBlue50</color>
 
-
+    <color name="colorofkyvlevelsbutton">#0B0AEE</color>
```
Comment (letelete) [inline mifospay/src/main/res/values/colors.xml]: Use camelCase

File: mifospay/src/main/res/values/dimens.xml [hunk 18]
```diff
@@ -30,24 +30,28 @@
 
     <dimen name="value_10dp">10dp</dimen>
     <dimen name="value_10sp">10sp</dimen>
+    <dimen name="value_16dp">16dp</dimen>
```
Comment (letelete) [inline mifospay/src/main/res/values/dimens.xml]: Don't declare values like that. All important margins are already declared so you should use these (if not add new margin value with proper name and not VALUE_16DP since it's just useless.

File: mifospay/src/main/res/values/strings.xml [hunk 19]
```diff
@@ -154,6 +154,7 @@
     <string name="cards">Cards</string>
     <string name="merchants">Merchants</string>
     <string name="kyc">KYC</string>
+    <string name="completekyc">Complete the KYC process</string>
```
Comment (letelete) [inline mifospay/src/main/res/values/strings.xml]: use snake_case

File: mifospay/src/main/res/layout/fragment_kyc.xml [hunk 20]
```diff
@@ -1,6 +1,181 @@
...
+
+            <LinearLayout
+                android:id="@+id/ll_lvl3"
+                android:layout_width="match_parent"
+                android:layout_height="match_parent"
+                android:background="@drawable/cardview_round_blue"
+                android:orientation="vertical"
+                android:padding="@dimen/value_10dp">
+
+                <TextView
```
Comment (letelete) [inline mifospay/src/main/res/layout/fragment_kyc.xml]: Did you just use CardViews instead of Buttons with 

style="@style/Widget.MaterialComponents.Button" or (...`Buton.OutlinedButton`)

I've told you about it earlier and you didn't even listen to my advices :|

**Comment by letelete** [comment]
Hey, @manddy! 

I'm glad about your contribution, but, as you might've noticed, there is a lot of things to fix in your pull request which are probable outcomes of your inexperience in computer programming. I recommend you give a try with some easier tasks, or just broaden your knowledge (by creating your own apps), and then come back to us!

Btw. That's not a material design icon.
You can find material icons here: https://material.io/tools/icons

Remember that we are using outlined theme for icons :)

So instead of this:

image

It should look like this:

image

File: mifospay/src/main/res/values/dimens.xml (hunk 18 above)
Comment (manddy) [reply mifospay/src/main/res/values/dimens.xml]: @letelete  I don't understand what is the problem in adding a dimen with value 16dp?kindly help me out with this

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java (hunk 6 above)
Comment (manddy) [reply mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: @letelete  sorry, did not understand where did I break the code style

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java (hunk 8 above)
Comment (manddy) [reply mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: @letelete  I think this code had to be changed for the kyc section to run properly kindly run and see it for yourself

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java (hunk 9 above)
Comment (manddy) [reply mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: ......

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java (hunk 10 above)
Comment (manddy) [reply mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: .....

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java (hunk 11 above)
Comment (manddy) [reply mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: @letelete  the llLvl1 had to be changed to white background  no other option

File: mifospay/src/main/res/layout/fragment_kyc.xml (hunk 20 above)
Comment (manddy) [reply mifospay/src/main/res/layout/fragment_kyc.xml]: @letelete  card views were used in the master branch of the app  so I used them in the redesign as well .What exactly do you want me to use ?

File: mifospay/src/main/res/values/dimens.xml (hunk 18 above)
Comment (letelete) [reply mifospay/src/main/res/values/dimens.xml]: Issues with this solution are:

1. Naming is just completely out-of-logic. What is the sense of having values named like that? It doesn't tell you what is this about, It doesn't tell where it has been used. 

So you could just set hard-coded values in the layout files as well and that would not make any difference. 

You should set names which actually tells something about the value.

We are keeping these ..."value_Xdp" dimens since they were used in the master branch, so to avoid any crashes we keep them here until all layout won't be redesigned.

2. Creating value with 16dp is actually redundant since you probably wanted to set it as a margin. We already have this value for margin and its called: marginItemsInSectionSmall

If this value stand for something else and not margin, you should add a new value but with the PROPER name. Like: `emptyStateImageSize`, because it tells you what is it used for.

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java (hunk 6 above)
Comment (letelete) [reply mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: Setting proper names to the fields in the classes are really important. Setting name like 't' is completely awful and it shouldn't have a place here. 

**One of the most important thing in the professional programming is to make code understandable for other developers.**

So instead of `R.id.third` it should look like `R.id.ll_level_3`

How would you know what does the `third` stands for after seeing it again like 3 weeks later? It's just a bad code.

Take a look at the other files to realize we are using name convention for the layout objects.

So instead of:

`t`

you should call it like:

`llLevel3`

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java (hunk 8 above)
Comment (letelete) [reply mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: You could make use of multiple on click assignment for the buttons (`@OnClick( { R.id.1, R.id.2, R.id.3 } )`) and then using a method with `View clickedButton` argument you could handle the view id in the switch to decide which fragment should've been opened.

File: mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java (hunk 11 above)
Comment (letelete) [reply mifospay/src/main/java/org/mifos/mobilewallet/mifospay/kyc/ui/KYCDescriptionFragment.java]: You should use a method which handles a proper argument to achieve the same effect but with x3 times less code.

DRY is also really important rule of computer programming. Take a look here:

* https://en.wikipedia.org/wiki/Don%27t_repeat_yourself
or here
* https://code.tutsplus.com/tutorials/3-key-software-principles-you-must-understand--net-25161
